### ISSUE TITLE
options to adjust format of paper background

### ISSUE THREAD

**Issue by fugata** [xournalpp/xournalpp#2137]
**Is your feature request related to a problem? Please describe.**

I use xournalpp to present lecture slides + take notes.  I often need to sketch graphs so the graph paper background is very helpful.  I have lots of equations with subscripts and subscripts, and the grid of the default graph paper is too small.

**Describe the solution you'd like**

The existing options for paper background are great -- it would be great if the user can tweak the parameters of the background.  E.g. for ruled paper: adjust width between the rules; for graph paper: specifiy the size of the grid.

Note that depending on the user input, the background patterns might not "determinate" evenly, i.e. the last row/column might be smaller the rest the regular ones.  The "correct" way to address that would have been to adjust the starting point of the background pattern to even things out.   But I have not problem having even "uneven" row/column at the end, and the flexibility of being able to tweak these parameters far exceed the small visual uneveness.

**Describe alternatives you've considered**

I manually created pdf graphpapers and use that as background.   But then I would not be able to change the background color, and I would not be able remove the background when I am done (my audience strongly prefer not to see the background patter during presentations).

**Additional context**

If it takes too much work to add this to the GUI interface, it's be perfectly fine for users to manually change the parameters in a config file  (this is likely to be something that a user would tweak once and for all and use it for all documents).

Thanks for the consideration!

**Comment by rolandlo** [comment]
In case of graph paper that should be pretty easy to implement, since the file format already supports it. For example the following xml file (gzipped and renamed from .gz to .xopp) produces a grid with size 50 pt (instead of 14.17 pt).
```

<?xml version="1.0" standalone="no"?>
<xournal creator="Xournal++ 1.1.0+dev" fileversion="4">
<title>Xournal++ document - see https://github.com/xournalpp/xournalpp</title>
<page width="595.27559100" height="841.88976400">
<background type="solid" color="#ffffffff" style="graph" config="r1=50.0"/>
<layer/>
</page>
</xournal>
```
As a workaround you may start with this xml-file (after changing the value of r1 in the xml-file and adding some more pages).

Your feature request definitively should be implemented. There is not much sense having a refined file format, when the gui does not support it.

**Comment by fugata** [comment]
> In case of graph paper that should be pretty easy to implement, since the file format already supports it. For example the following xml file (gzipped and renamed from .gz to .xopp) produces a grid with size 50 pt (instead of 14.17 pt).
> 
> ```
> 
> <?xml version="1.0" standalone="no"?>
> <xournal creator="Xournal++ 1.1.0+dev" fileversion="4">
> <title>Xournal++ document - see https://github.com/xournalpp/xournalpp</title>
> <page width="595.27559100" height="841.88976400">
> <background type="solid" color="#ffffffff" style="graph" config="r1=50.0"/>
> <layer/>
> </page>
> </xournal>
> ```
> 
> As a workaround you may start with this xml-file (after changing the value of r1 in the xml-file and adding some more pages).
> 
> Your feature request definitively should be implemented. There is not much sense having a refined file format, when the gui does not support it.

Thanks for your prompt reply and consideration!  Where should I put this .xopp file?

**Comment by rolandlo** [comment]
The xml file is the one you see above (human readable). A xopp-file is a compressed xml-file (not human readable any more). So you would have to
1. Copy the content of the above file into a text editor and save it, say under filename "backgroundTemplate", with no file extension. It does not matter where you save it. You should just find it easily.
2. Modify the "r1=50.0" on the 5th line according to your needs (the number stands for the grid size). 
3. gzip the file (for instance by running `gzip backgroundTemplate` in the terminal in this folder, assuming you are on Linux; or by right-clicking and choosing "compress" in the context-menu). This should produce the file "backgroundTemplate.gz"
4. Rename the file to "backgroundTemplate.xopp". 

Now whenever you want to have a new document with your new (larger) grid size just open the backgroundTemplate.xopp file in Xournal++, save it under a new name. If you add a page, it will automatically have the same graph background (same grid size).

To learn more about the file format you could consult the [Xournal User's manual](http://www-math.mit.edu/~auroux/software/xournal/manual.html#file-format). It describes the .xoj-format of Xournal, which is a subset of the Xournal++ .xopp-format.

If you need more help, let me know.

**Comment by fugata** [comment]
> The xml file is the one you see above (human readable). A xopp-file is a compressed xml-file (not human readable any more). So you would have to
> 
>     1. Copy the content of the above file into a text editor and save it, say under filename "backgroundTemplate", with no file extension. It does not matter where you save it. You should just find it easily.
> 
>     2. Modify the "r1=50.0" on the 5th line according to your needs (the number stands for the grid size).
> 
>     3. gzip the file (for instance by running `gzip backgroundTemplate` in the terminal in this folder, assuming you are on Linux; or by right-clicking and choosing "compress" in the context-menu). This should produce the file "backgroundTemplate.gz"
> 
>     4. Rename the file to "backgroundTemplate.xopp".
> 
> 
> Now whenever you want to have a new document with your new (larger) grid size just open the backgroundTemplate.xopp file in Xournal++, save it under a new name. If you add a page, it will automatically have the same graph background (same grid size).
> 
> To learn more about the file format you could consult the [Xournal User's manual](http://www-math.mit.edu/~auroux/software/xournal/manual.html#file-format). It describes the .xoj-format of Xournal, which is a subset of the Xournal++ .xopp-format.
> 
> If you need more help, let me know.

This works beatifully.  Thanks SO MUCH for your help!

**Comment by rolandlo** [comment]
There is actually a simpler (and better) way to adjust the format of the paper background. That's what you need to do:

1. Find the relevant pagetemplates.ini file (in the latest Linux version the file is located in `/usr/share/xournalpp/ui/`)
2. Open the file (with writing permissions) and add a new entry like
```
[largeGraph]
name=Graph with large grid
name[de]=Kariert mit grossem Gitter
format=graph
config=r1=50

```
3. Restart Xournal++ and check whether Xournal++ has registered the name page template (under `journal->page background`) or using the toolbar (next to the "add page" icon). 
4. In case Xournal++ has not added the new page template, then most likely you have several `pagetemplates.ini` files on the system and you must try another one.

**Comment by Tydblair** [comment]
> There is actually a simpler (and better) way to adjust the format of the paper background. That's what you need to do:
> 
> 1. Find the relevant pagetemplates.ini file (in the latest Linux version of Xournal++ the file is located in `/usr/share/xournalpp/ui/`)
> 2. Open the file (with writing permissions) and add a new entry like
> 
> ```
> [largeGraph]
> name=Graph with large grid
> name[de]=Kariert mit grossem Gitter
> format=graph
> config=r1=50
> ```
> 
> 1. Restart Xournal++ and check whether Xournal++ has registered the name page template (under `journal->page background`) or using the toolbar (next to the "add page" icon).
> 2. In case Xournal++ has not added the new page template, then most likely you have several `pagetemplates.ini` files on the system and you must try another one.

Are you able to help me understand the dimentions of these parametres? 

im guessing that
r1 = Ruled line dimentions
m1 = Margin dimentions
F1 = Colour in Hex

But I dont understand their relationship to mm, can you point me in the right direction so that I can create some custom templates

**Comment by rolandlo** [comment]
> 
> Are you able to help me understand the dimentions of these parametres?
> 
> im guessing that
> r1 = Ruled line dimentions
> m1 = Margin dimentions
> F1 = Colour in Hex
> 
> But I dont understand their relationship to mm, can you point me in the right direction so that I can create some custom templates

First of all, the unit is 1 pt = 1/72 inch. So 5mm = 14.17 pt is the default grid size of the graph paper.
`r1` = grid size (for graph and dotted paper backgrounds)
`m1` = margin (left, right, top and bottom)
`f1`= primary foreground color (in Hex)

There are also
`lw` = line width,
`rm` = boolean with values 0 (false) or 1 (true), if rounding should take place for the margin, so you will have an integral number of grid lines
`f2`= secondary foreground color (in Hex)

**Comment by prbuen** [comment]
Is there a config format for ruled backgrounds (format=lined)?

**Comment by d1payan** [comment]
> First of all, the unit is 1 pt = 1/72 inch. So 5mm = 14.17 pt is the default grid size of the graph paper. `r1` = grid size (for graph and dotted paper backgrounds) `m1` = margin (left, right, top and bottom) `f1`= primary foreground color (in Hex)
> 
> There are also `lw` = line width, `rm` = boolean with values 0 (false) or 1 (true), if rounding should take place for the margin, so you will have an integral number of grid lines `f2`= secondary foreground color (in Hex) `af1`= alternative secondary foreground color (in Hex) when page background is dark `af2`= alternative secondary foreground color (in Hex) when page background is dark
> 
> **Edit:** Added `af1` and `af2`.



> 

What is the command for increasing gap between ruled lines ???

**Comment by rolandlo** [comment]
> 
> What is the command for increasing gap between ruled lines ???

The gap between ruled lines can only be adjusted since May 2022 (not yet in the stable version). Use the parameter `r1`. I have updated the description.

**Comment by UnknownPerson-0x00** [comment]
Where is the `/usr/share/xournalpp/ui/pagetemplates.ini` file located in the flatpak release of xournalpp?

**Comment by i-zanis** [comment]
I am testing the following for Ruled format but nothing changes.

[test2]
name=test2
name[de]=Liniert
format=ruled
config=r1=1

**Comment by rolandlo** [comment]
@i-zanis What you added works for me, although `r1=1` is smaller than you want (it's 1pt = 1/72 inch). Do you use the [nightly version](https://github.com/xournalpp/xournalpp/releases/tag/nightly)? It won't work in the stable version. How do you apply the new page background?  When you choose it in the dropdown menu from the toolbar 
![grafik](https://user-images.githubusercontent.com/40485433/207502880-b90bdb28-a977-4e82-9a23-cd0ee2f0f350.png)
you have to add a new page with that format (or open the dropdown again and click "apply to current page") to get it.

**Comment by NNoah18** [comment]
> m1

I am able to change the right vertical line placement for ruled paper using m1=36, but how would I change the margins of the top and bottom as well?

**Comment by NNoah18** [comment]
> > Are you able to help me understand the dimentions of these parametres?
> > im guessing that
> > r1 = Ruled line dimentions
> > m1 = Margin dimentions
> > F1 = Colour in Hex
> > But I dont understand their relationship to mm, can you point me in the right direction so that I can create some custom templates
> 
> First of all, the unit is 1 pt = 1/72 inch. So 5mm = 14.17 pt is the default grid size of the graph paper. `r1` = grid size (for graph and dotted paper backgrounds) `m1` = margin (left, right, top and bottom) `f1`= primary foreground color (in Hex)
> 
> There are also `lw` = line width, `rm` = boolean with values 0 (false) or 1 (true), if rounding should take place for the margin, so you will have an integral number of grid lines `f2`= secondary foreground color (in Hex) `af1`= alternative secondary foreground color (in Hex) when page background is dark `af2`= alternative secondary foreground color (in Hex) when page background is dark
> 
> The following two parameters for ruled/lined backgrounds are only available in the development version (nightly release), not yet in the stable version 1.1.1.
> 
> For ruled and lined paper there is the parameter `r1` = spacing between horizontal lines
> 
> For lined paper there is also the parameter `m1` = margin for the vertical line (positive values for a vertical lines on the left, negative values for a vertical line on the right)
> 
> For isometric paper there is the parameter `r1` = side length of the triangles
> 
> **Edit:** Added `af1` and `af2`. **Edit2:** Added `m1` and `r1` in case of ruled/lined paper (see also the [description](https://github.com/xournalpp/xournalpp/pull/3969#issue-1208334299) of PR #3969) **Edit3:** Added `r1` in case of isometric paper

I am able to change the right vertical line placement for ruled paper using m1=36, but how would I change the margins of the top and bottom as well?

**Comment by NNoah18** [comment]
> > Are you able to help me understand the dimentions of these parametres?
> > im guessing that
> > r1 = Ruled line dimentions
> > m1 = Margin dimentions
> > F1 = Colour in Hex
> > But I dont understand their relationship to mm, can you point me in the right direction so that I can create some custom templates
> 
> First of all, the unit is 1 pt = 1/72 inch. So 5mm = 14.17 pt is the default grid size of the graph paper. `r1` = grid size (for graph and dotted paper backgrounds) `m1` = margin (left, right, top and bottom) `f1`= primary foreground color (in Hex)
> 
> There are also `lw` = line width, `rm` = boolean with values 0 (false) or 1 (true), if rounding should take place for the margin, so you will have an integral number of grid lines `f2`= secondary foreground color (in Hex) `af1`= alternative secondary foreground color (in Hex) when page background is dark `af2`= alternative secondary foreground color (in Hex) when page background is dark
> 
> The following two parameters for ruled/lined backgrounds are only available in the development version (nightly release), not yet in the stable version 1.1.1.
> 
> For ruled and lined paper there is the parameter `r1` = spacing between horizontal lines
> 
> For lined paper there is also the parameter `m1` = margin for the vertical line (positive values for a vertical lines on the left, negative values for a vertical line on the right)
> 
> For isometric paper there is the parameter `r1` = side length of the triangles
> 
> **Edit:** Added `af1` and `af2`. **Edit2:** Added `m1` and `r1` in case of ruled/lined paper (see also the [description](https://github.com/xournalpp/xournalpp/pull/3969#issue-1208334299) of PR #3969) **Edit3:** Added `r1` in case of isometric paper

@rolandlo 
I am able to change the vertical line placement for ruled paper using m1=36, but how would I also change the top and bottom margin dimensions? - I want the ruled lines to start sooner on the page and end later on the page

**Comment by bhennion** [comment]
> @rolandlo I am able to change the vertical line placement for ruled paper using m1=36, but how would I also change the top and bottom margin dimensions? - I want the ruled lines to start sooner on the page and end later on the page

Top and bottom margins are currently not configurable. They are hardcoded values.

**Comment by christianhorne** [comment]
is it possible to adjust the line-width for rectangles on a graph background?

**Comment by Simbi0z** [comment]
Is there any chance for a GUI for modifying these settings?
Thanks.

**Comment by Ainson** [comment]
Thanks for the info. I also wanted a bigger grid. Since Xournal snap to grid is based on 5mm = 14.17 I made the grid double 28,24 and that works fine.

I also wanted the grid to show up on a darker background so added the color code af1=0x00ff00 for light green. (af1= alternative secondary foreground color)

So this is what my new section looks like:

[largeGraph]
name=Graph with large grid
format=graph
config=r1=28.34,af1=0x00ff00


NOTE: I first got an error when loading Xournal after editing pagetemplate.ini, that turned out was because of using Windows Notepad. (error: cannot read pagetemplate.ini)

The solution was to use SciTE, a dedicated program language editor, so it seems normal text editors screws up the file. (there are other free ones like Scintilla, Visual Studio Code etc).
